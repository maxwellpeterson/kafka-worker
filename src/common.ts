export interface Env {
  // The hostname of this worker
  HOSTNAME: string;
  CLUSTER: DurableObjectNamespace;
  // Example binding to KV. Learn more at https://developers.cloudflare.com/workers/runtime-apis/kv/
  // MY_KV_NAMESPACE: KVNamespace;
  //
  // Example binding to R2. Learn more at https://developers.cloudflare.com/workers/runtime-apis/r2/
  // MY_BUCKET: R2Bucket;
}

export type ValueOf<T> = T[keyof T];
