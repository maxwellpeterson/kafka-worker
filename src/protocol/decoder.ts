import {
  Int16,
  Int32,
  int16Size,
  int32Size,
  String,
  KafkaArray,
  ErrorCode,
  validErrorCode,
  Int64,
  int64Size,
} from "src/protocol/common";

export class Decoder {
  private readonly view: DataView;
  private offset: number;

  constructor(buffer: ArrayBuffer) {
    this.view = new DataView(buffer);
    this.offset = 0;
  }

  readInt16(): Int16 {
    const value = this.view.getInt16(this.offset);
    this.offset += int16Size;
    return value;
  }

  readEnum<T extends Int16>(predicate: (value: Int16) => value is T): T {
    const value = this.readInt16();
    if (!predicate(value)) {
      throw new Error(`Invalid enum value: ${value}`);
    }
    return value;
  }

  readInt32(): Int32 {
    const value = this.view.getInt32(this.offset);
    this.offset += int32Size;
    return value;
  }

  readInt64(): Int64 {
    const value = this.view.getBigInt64(this.offset);
    this.offset += int64Size;
    return value;
  }

  readString(): String {
    const size = this.readInt16();
    if (size === -1) {
      // I don't think this should ever happen...
      throw new Error("Unexpected null string!");
    }
    const stringBuffer = this.view.buffer.slice(
      this.offset,
      this.offset + size
    );
    const value = new TextDecoder().decode(stringBuffer);
    this.offset += size;
    return value;
  }

  readKafkaArray<T>(readElement: (index: number) => T): KafkaArray<T> {
    const size = this.readInt32();
    if (size === -1) {
      return null;
    }
    const values = new Array<T>(size);
    for (let i = 0; i < size; i++) {
      values[i] = readElement(i);
    }
    return values;
  }

  // In several apis, there is no semantic difference between empty and null
  // arrays, and converting null arrays to empty arrays eliminates unnecessary
  // code pollution from null checking
  readArray<T>(readElement: (index: number) => T): Array<T> {
    return this.readKafkaArray(readElement) ?? [];
  }

  readMessageSet(size: Int32): ArrayBuffer {
    const buffer = this.view.buffer.slice(this.offset, this.offset + size);
    this.offset += size;
    return buffer;
  }
}

export class KafkaRequestDecoder extends Decoder {
  constructor(buffer: ArrayBuffer) {
    super(buffer);

    const expectedSize = this.readInt32();
    const actualSize = buffer.byteLength - int32Size;

    if (expectedSize !== actualSize) {
      throw new Error(
        `Message length does not match size header: expected ${expectedSize} but got ${actualSize}`
      );
    }
  }
}
