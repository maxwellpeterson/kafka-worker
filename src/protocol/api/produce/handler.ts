import { Env } from "src/common";
import {
  ProduceRequest,
  ProduceResponse,
} from "src/protocol/api/produce/types";

export const handleProduceRequest = async (
  env: Env,
  request: ProduceRequest
): Promise<ProduceResponse | null> => {
  return null;
};
