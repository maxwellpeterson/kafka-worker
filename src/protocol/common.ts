import { ValueOf } from "src/common";

// These don't add any type safety, but help with labeling
export type Int16 = number;
export type Int32 = number;
export type Int64 = bigint;

export const int16Size = 2;
export const int32Size = 4;
export const int64Size = 8;

export type KafkaArray<T> = T[] | null;

export interface RequestMetadata {
  apiVersion: Int16;
  correlationId: Int32;
  clientId: string;
}
export const generateEnumPredicate =
  <E extends Record<string, V>, V>(enumObj: E) =>
  (value: V): value is ValueOf<E> =>
    Object.values<V>(enumObj).includes(value);

// ApiKey is an Int16
export const ApiKey = {
  Produce: 0,
  // Fetch: 1,
  // ListOffsets: 2,
  Metadata: 3,
} as const;
export type ApiKey = ValueOf<typeof ApiKey>;
export const validApiKey = generateEnumPredicate(ApiKey);

// ErrorCode is an Int16
export const ErrorCode = {
  None: 0,
  UnknownTopicOrPartition: 3,
} as const;
export type ErrorCode = ValueOf<typeof ErrorCode>;
export const validErrorCode = generateEnumPredicate(ErrorCode);

// Acks is an Int16
export const Acks = {
  None: 0,
  Leader: 1,
  FullISR: -1,
} as const;
export type Acks = ValueOf<typeof Acks>;
export const validAcks = generateEnumPredicate(Acks);
