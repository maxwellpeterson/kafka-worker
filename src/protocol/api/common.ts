import { Env } from "src/common";
import { Decoder } from "src/protocol/decoder";
import { Encoder } from "src/protocol/encoder";

export type EntrypointFn<Output> = (
  env: Env,
  version: number,
  decoder: Decoder,
  encoder: Encoder
) => Promise<Output>;

export type CheckVersionFn = (version: number) => void;
export type DecodeFn<Req> = (decoder: Decoder) => Req;
export type HandleFn<Req, Res> = (env: Env, request: Req) => Promise<Res>;
export type EncodeFn<Res, Output> = (encoder: Encoder, response: Res) => Output;

export const generateEntrypointFn =
  <Req, Res, Output>(
    apiName: string,
    checkVersion: CheckVersionFn,
    decode: DecodeFn<Req>,
    handle: HandleFn<Req, Res>,
    encode: EncodeFn<Res, Output>
  ): EntrypointFn<Output> =>
  async (
    env: Env,
    version: number,
    decoder: Decoder,
    encoder: Encoder
  ): Promise<Output> => {
    checkVersion(version);

    const request = decode(decoder);
    console.log(
      `Received ${apiName} request: ${JSON.stringify(request, null, 2)}`
    );

    const response = await handle(env, request);
    console.log(
      `Sending ${apiName} response: ${JSON.stringify(response, null, 2)}`
    );

    return encode(encoder, response);
  };

export const generateCheckVersionFn =
  (apiName: string, expected: number): CheckVersionFn =>
  (version: number) => {
    if (version !== expected) {
      throw new Error(
        `Unsupported version of ${apiName} api: expected ${expected} but got ${version}`
      );
    }
  };
