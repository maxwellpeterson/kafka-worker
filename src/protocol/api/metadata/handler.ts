import { Env } from "src/common";
import {
  MetadataRequest,
  MetadataResponse,
} from "src/protocol/api/metadata/types";
import { KafkaArray, KafkaString } from "src/protocol/common";
import { fetchClusterMetadata } from "src/state/cluster";

export const handleMetadataRequest = async (
  env: Env,
  request: MetadataRequest
): Promise<MetadataResponse> => {
  return await fetchClusterMetadata(env, cleanTopics(request.topics));
};

const cleanTopics = (topics: KafkaArray<KafkaString>): string[] => {
  if (topics === null) {
    return [];
  }
  return topics.filter((topic): topic is string => topic !== null);
};
