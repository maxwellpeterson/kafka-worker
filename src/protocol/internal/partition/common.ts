import { ValueOf } from "src/common";
import {
  ApiKey,
  Int16,
  Int32,
  generateEnumPredicate,
  int32Size,
} from "src/protocol/common";
import { Decoder } from "src/protocol/decoder";
import { Encoder } from "src/protocol/encoder";

// Partition Request (Version: 0) => header request_message
//   header => api_key api_version correlation_id client_id
//     api_key => INT16
//     api_version => INT16
//     correlation_id => INT32
//     client_id => STRING
//   request_message => PARTITION_PRODUCE_REQUEST

export interface PartitionRequestHeader {
  apiKey: PartitionApiKey;
  apiVersion: Int16;
  correlationId: Int32;
  clientId: string;
}

export const encodePartitionRequestHeader = (
  encoder: Encoder,
  header: PartitionRequestHeader
) => {
  encoder.writeEnum(header.apiKey);
  encoder.writeInt16(header.apiVersion);
  encoder.writeInt32(header.correlationId);
  encoder.writeString(header.clientId);
};

export const decodePartitionRequestHeader = (
  decoder: Decoder
): PartitionRequestHeader => {
  return {
    apiKey: decoder.readEnum(validPartitionApiKey),
    apiVersion: decoder.readInt16(),
    correlationId: decoder.readInt32(),
    clientId: decoder.readString(),
  };
};

// PartitionApiKey is an Int16
export const PartitionApiKey = {
  Produce: ApiKey.Produce,
} as const;
export type PartitionApiKey = ValueOf<typeof PartitionApiKey>;
const validPartitionApiKey = generateEnumPredicate(PartitionApiKey);

// Partition Response (Version: 0) => correlation_id response_message
//   correlation_id => INT32
//   response_message => PARTITION_PRODUCE_RESPONSE

export class PartitionResponseEncoder extends Encoder {
  constructor(correlationId: Int32, initialBufferSize = 64) {
    super(initialBufferSize + int32Size);
    this.writeInt32(correlationId);
  }
}
