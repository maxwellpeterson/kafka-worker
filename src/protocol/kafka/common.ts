import { Int32, int32Size } from "src/protocol/common";
import { Decoder } from "src/protocol/decoder";
import { Encoder } from "src/protocol/encoder";

export class KafkaResponseEncoder extends Encoder {
  constructor(correlationId: Int32, initialBufferSize = 64) {
    super(initialBufferSize + 2 * int32Size);
    // Reserve space for size header at front of buffer
    this.offset += int32Size;
    this.writeInt32(correlationId);
  }

  // Overrides parent method, adds size header to buffer
  buffer(): ArrayBuffer {
    // Write size header to reserved space at front of buffer
    this.view.setInt32(0, this.offset - int32Size);
    return super.buffer();
  }
}

export class KafkaRequestDecoder extends Decoder {
  constructor(buffer: ArrayBuffer) {
    super(buffer);

    const expectedSize = this.readInt32();
    const actualSize = buffer.byteLength - int32Size;

    if (expectedSize !== actualSize) {
      throw new Error(
        `Message length does not match size header: expected ${expectedSize} but got ${actualSize}`
      );
    }
  }
}
