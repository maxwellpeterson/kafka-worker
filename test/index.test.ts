import handler from "src/index";
import { Acks, ApiKey, Int32 } from "src/protocol/common";
import { Decoder } from "src/protocol/decoder";
import { KafkaDecoder, KafkaRequestEncoder } from "src/protocol/kafka/common";
import {
  KafkaFetchRequest,
  KafkaFetchResponse,
  decodeKafkaFetchResponse,
  encodeKafkaFetchRequest,
} from "src/protocol/kafka/fetch";
import {
  KafkaListOffsetsRequest,
  KafkaListOffsetsResponse,
  decodeKafkaListOffsetsResponse,
  encodeKafkaListOffsetsRequest,
} from "src/protocol/kafka/list-offsets";
import {
  KafkaMetadataResponse,
  decodeKafkaMetadataResponse,
  encodeKafkaMetadataRequest,
} from "src/protocol/kafka/metadata";
import {
  KafkaProduceRequest,
  KafkaProduceResponse,
  decodeKafkaProduceResponse,
  encodeKafkaProduceRequest,
} from "src/protocol/kafka/produce";
import { base64, fillMessageSet } from "test/common";

class GatewayConn {
  private readonly socket: WebSocket;
  private readonly unhandled: (ArrayBuffer | string)[];
  private handleMessage?: (message: ArrayBuffer) => void;

  constructor() {
    const env = getMiniflareBindings();

    const response = handler.fetch(
      new Request("https://fetch.handler", {
        headers: { Upgrade: "websocket" },
      }),
      env
    );
    if (!response.ok) {
      throw new Error(`Initial request failed with status ${response.status}`);
    }

    const socket = response.webSocket;
    if (socket === null) {
      throw new Error("Initial response does not include WebSocket");
    }

    socket.accept();
    this.socket = socket;
    this.unhandled = [];

    this.socket.addEventListener("message", (event) => {
      if (typeof event.data === "string" || this.handleMessage === undefined) {
        this.unhandled.push(event.data);
        return;
      }
      this.handleMessage(event.data);
    });
  }

  request(buffer: ArrayBuffer): Promise<ArrayBuffer> {
    if (this.unhandled.length > 0) {
      throw new Error(
        `Can't send next message when there are ${this.unhandled.length} unhandled messages`
      );
    }

    if (this.handleMessage !== undefined) {
      throw new Error(
        "Can't send next message when there is a request in flight"
      );
    }

    this.socket.send(buffer);

    return new Promise((resolve) => {
      this.handleMessage = (message: ArrayBuffer) => {
        this.handleMessage = undefined;
        resolve(message);
      };
    });
  }

  close() {
    if (this.unhandled.length > 0) {
      throw new Error(
        `Can't close connection when there are ${this.unhandled.length} unhandled messages`
      );
    }

    this.socket.close();
  }
}

type DecodeFunc<T> = (decoder: Decoder) => T;
type RequestResponse<T> = [Int32, ArrayBuffer, DecodeFunc<T>];
type TestCase<T> = [string, string, RequestResponse<T>[]];

const makeMetadataPair = (
  correlationId: Int32,
  topics: string[]
): RequestResponse<KafkaMetadataResponse> => {
  const encoder = new KafkaRequestEncoder({
    apiKey: ApiKey.Metadata,
    apiVersion: 0,
    correlationId,
    clientId: null,
  });
  const request = encodeKafkaMetadataRequest(encoder, {
    topics,
  });
  return [correlationId, request, decodeKafkaMetadataResponse];
};

const makeProducePair = (
  correlationId: Int32,
  topics: KafkaProduceRequest["topics"]
): RequestResponse<KafkaProduceResponse> => {
  const encoder = new KafkaRequestEncoder({
    apiKey: ApiKey.Produce,
    apiVersion: 0,
    correlationId,
    clientId: null,
  });
  const request = encodeKafkaProduceRequest(encoder, {
    acks: Acks.Leader,
    timeoutMs: 10_000,
    topics,
  });
  return [correlationId, request, decodeKafkaProduceResponse];
};

const makeFetchPair = (
  correlationId: Int32,
  request: KafkaFetchRequest
): RequestResponse<KafkaFetchResponse> => {
  const encoder = new KafkaRequestEncoder({
    apiKey: ApiKey.Fetch,
    apiVersion: 0,
    correlationId,
    clientId: null,
  });
  const encoded = encodeKafkaFetchRequest(encoder, request);
  return [correlationId, encoded, decodeKafkaFetchResponse];
};

const makeListOffsetsPair = (
  correlationId: Int32,
  request: KafkaListOffsetsRequest
): RequestResponse<KafkaListOffsetsResponse> => {
  const encoder = new KafkaRequestEncoder({
    apiKey: ApiKey.ListOffsets,
    apiVersion: 0,
    correlationId,
    clientId: null,
  });
  const encoded = encodeKafkaListOffsetsRequest(encoder, request);
  return [correlationId, encoded, decodeKafkaListOffsetsResponse];
};

describe("Kafka API", () => {
  const cases: TestCase<
    | KafkaMetadataResponse
    | KafkaProduceResponse
    | KafkaListOffsetsResponse
    | KafkaFetchResponse
  >[] = [
    ["metadata", "fetch all topics", [makeMetadataPair(5, [])]],
    ["metadata", "fetch specific topic", [makeMetadataPair(5, ["test-topic"])]],
    [
      "metadata",
      "fetch nonexistent topic",
      [makeMetadataPair(5, ["other-topic"])],
    ],
    [
      "metadata",
      "fetch specific topic and nonexistent topic",
      [makeMetadataPair(5, ["test-topic", "other-topic"])],
    ],
    [
      "produce",
      "send one message to one partition",
      [
        makeProducePair(5, [
          {
            name: "test-topic",
            partitions: [{ index: 333, messageSet: fillMessageSet(1) }],
          },
        ]),
      ],
    ],
    [
      "produce",
      "send multiple message batches to one partition",
      [
        makeProducePair(5, [
          {
            name: "other-topic",
            partitions: [{ index: 999, messageSet: fillMessageSet(3) }],
          },
        ]),
        makeProducePair(6, [
          {
            name: "other-topic",
            partitions: [{ index: 999, messageSet: fillMessageSet(2) }],
          },
        ]),
      ],
    ],
    [
      "listOffsets",
      "initial offsets",
      [
        makeListOffsetsPair(0, {
          replicaId: -1,
          topics: [
            {
              name: "test-topic",
              partitions: [
                { index: 0, timestamp: BigInt(0), maxNumOffsets: 16 },
              ],
            },
          ],
        }),
      ],
    ],
    [
      "listOffsets",
      "full offset list after producing records",
      [
        makeProducePair(0, [
          {
            name: "test-topic",
            partitions: [{ index: 0, messageSet: fillMessageSet(10) }],
          },
        ]),
        makeListOffsetsPair(1, {
          replicaId: -1,
          topics: [
            {
              name: "test-topic",
              partitions: [
                { index: 0, timestamp: BigInt(0), maxNumOffsets: 16 },
              ],
            },
          ],
        }),
      ],
    ],
    [
      "listOffsets",
      "oldest offset after producing records",
      [
        makeProducePair(0, [
          {
            name: "test-topic",
            partitions: [{ index: 0, messageSet: fillMessageSet(10) }],
          },
        ]),
        makeListOffsetsPair(1, {
          replicaId: -1,
          topics: [
            {
              name: "test-topic",
              partitions: [
                { index: 0, timestamp: BigInt(-2), maxNumOffsets: 1 },
              ],
            },
          ],
        }),
      ],
    ],
    [
      "listOffsets",
      "maximum number of offsets is respected",
      [
        makeProducePair(0, [
          {
            name: "test-topic",
            partitions: [{ index: 0, messageSet: fillMessageSet(10) }],
          },
        ]),
        makeListOffsetsPair(1, {
          replicaId: -1,
          topics: [
            {
              name: "test-topic",
              partitions: [
                { index: 0, timestamp: BigInt(0), maxNumOffsets: 4 },
              ],
            },
          ],
        }),
      ],
    ],
    [
      "fetch",
      "fetch on empty partition",
      [
        makeFetchPair(0, {
          replicaId: -1,
          maxWaitMs: 500,
          minBytes: 256,
          topics: [
            {
              name: "test-topic",
              partitions: [
                { index: 0, fetchOffset: BigInt(0), maxBytes: 8192 },
              ],
            },
          ],
        }),
      ],
    ],
  ];

  test.each(cases)("%s: %s", async (_api, _name, pairs) => {
    const gateway = new GatewayConn();

    for (const [correlationId, request, decodeResponse] of pairs) {
      const response = await gateway.request(request);
      expect(base64(response)).toMatchSnapshot();

      const decoder = new KafkaDecoder(response);
      expect(decoder.readInt32()).toEqual(correlationId);

      const decoded = decodeResponse(decoder);
      expect(decoded).toMatchSnapshot();
    }

    gateway.close();
  });

  test("produce message set then fetch message set", async () => {
    const gateway = new GatewayConn();
    const messageSet = fillMessageSet(5);

    const produceCorrelationId = 0;
    const produceEncoder = new KafkaRequestEncoder({
      apiKey: ApiKey.Produce,
      apiVersion: 0,
      correlationId: produceCorrelationId,
      clientId: null,
    });
    const produceRequest = encodeKafkaProduceRequest(produceEncoder, {
      acks: Acks.Leader,
      timeoutMs: 10_000,
      topics: [{ name: "test-topic", partitions: [{ index: 0, messageSet }] }],
    });
    const produceResponse = await gateway.request(produceRequest);
    const produceDecoder = new KafkaDecoder(produceResponse);
    expect(produceDecoder.readInt32()).toEqual(produceCorrelationId);
    expect(decodeKafkaProduceResponse(produceDecoder)).toMatchInlineSnapshot(`
      Object {
        "topics": Array [
          Object {
            "name": "test-topic",
            "partitions": Array [
              Object {
                "baseOffset": 0n,
                "errorCode": 0,
                "index": 0,
              },
            ],
          },
        ],
      }
    `);

    const fetchCorrelationId = 1;
    const fetchEncoder = new KafkaRequestEncoder({
      apiKey: ApiKey.Fetch,
      apiVersion: 0,
      correlationId: fetchCorrelationId,
      clientId: null,
    });
    const fetchRequest = encodeKafkaFetchRequest(fetchEncoder, {
      replicaId: -1,
      maxWaitMs: 1000,
      minBytes: 64,
      topics: [
        {
          name: "test-topic",
          partitions: [{ index: 0, fetchOffset: BigInt(0), maxBytes: 8192 }],
        },
      ],
    });
    const fetchResponse = await gateway.request(fetchRequest);
    const fetchDecoder = new KafkaDecoder(fetchResponse);
    expect(fetchDecoder.readInt32()).toEqual(fetchCorrelationId);
    const fetchResponseDecoded = decodeKafkaFetchResponse(fetchDecoder);
    expect(fetchResponseDecoded.topics[0].partitions[0].messageSet).toEqual(
      messageSet
    );
    // Exclude buffer data from inline snapshot (because it's huge)
    fetchResponseDecoded.topics[0].partitions[0].messageSet = new Uint8Array();
    expect(fetchResponseDecoded).toMatchInlineSnapshot(`
      Object {
        "topics": Array [
          Object {
            "name": "test-topic",
            "partitions": Array [
              Object {
                "errorCode": 0,
                "highWatermark": 5n,
                "index": 0,
                "messageSet": Uint8Array [],
              },
            ],
          },
        ],
      }
    `);
  });
});
