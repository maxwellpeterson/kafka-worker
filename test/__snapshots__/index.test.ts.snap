// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kafka API metadata: fetch all topics 1`] = `"AAAAQwAAAAUAAAABAAABTQAJbG9jYWxob3N0AAAiUwAAAAEAAAAKdGVzdC10b3BpYwAAAAEAAAAAAAAAAAFNAAAAAAAAAAA="`;

exports[`Kafka API metadata: fetch all topics 2`] = `
Object {
  "brokers": Array [
    Object {
      "host": "localhost",
      "nodeId": 333,
      "port": 8787,
    },
  ],
  "topics": Array [
    Object {
      "errorCode": 0,
      "name": "test-topic",
      "partitions": Array [
        Object {
          "errorCode": 0,
          "isrNodes": Array [],
          "leaderId": 333,
          "partitionIndex": 0,
          "replicaNodes": Array [],
        },
      ],
    },
  ],
}
`;

exports[`Kafka API metadata: fetch nonexistent topic 1`] = `"AAAAMgAAAAUAAAABAAABTQAJbG9jYWxob3N0AAAiUwAAAAEAAwALb3RoZXItdG9waWMAAAAA"`;

exports[`Kafka API metadata: fetch nonexistent topic 2`] = `
Object {
  "brokers": Array [
    Object {
      "host": "localhost",
      "nodeId": 333,
      "port": 8787,
    },
  ],
  "topics": Array [
    Object {
      "errorCode": 3,
      "name": "other-topic",
      "partitions": Array [],
    },
  ],
}
`;

exports[`Kafka API metadata: fetch specific topic 1`] = `"AAAAQwAAAAUAAAABAAABTQAJbG9jYWxob3N0AAAiUwAAAAEAAAAKdGVzdC10b3BpYwAAAAEAAAAAAAAAAAFNAAAAAAAAAAA="`;

exports[`Kafka API metadata: fetch specific topic 2`] = `
Object {
  "brokers": Array [
    Object {
      "host": "localhost",
      "nodeId": 333,
      "port": 8787,
    },
  ],
  "topics": Array [
    Object {
      "errorCode": 0,
      "name": "test-topic",
      "partitions": Array [
        Object {
          "errorCode": 0,
          "isrNodes": Array [],
          "leaderId": 333,
          "partitionIndex": 0,
          "replicaNodes": Array [],
        },
      ],
    },
  ],
}
`;

exports[`Kafka API metadata: fetch specific topic and nonexistent topic 1`] = `"AAAAVgAAAAUAAAABAAABTQAJbG9jYWxob3N0AAAiUwAAAAIAAAAKdGVzdC10b3BpYwAAAAEAAAAAAAAAAAFNAAAAAAAAAAAAAwALb3RoZXItdG9waWMAAAAA"`;

exports[`Kafka API metadata: fetch specific topic and nonexistent topic 2`] = `
Object {
  "brokers": Array [
    Object {
      "host": "localhost",
      "nodeId": 333,
      "port": 8787,
    },
  ],
  "topics": Array [
    Object {
      "errorCode": 0,
      "name": "test-topic",
      "partitions": Array [
        Object {
          "errorCode": 0,
          "isrNodes": Array [],
          "leaderId": 333,
          "partitionIndex": 0,
          "replicaNodes": Array [],
        },
      ],
    },
    Object {
      "errorCode": 3,
      "name": "other-topic",
      "partitions": Array [],
    },
  ],
}
`;

exports[`Kafka API produce: send multiple message batches to one partiton 1`] = `"AAAAJwAAAAUAAAABAAtvdGhlci10b3BpYwAAAAEAAAPnAAAAAAAAAAAAAA=="`;

exports[`Kafka API produce: send multiple message batches to one partiton 2`] = `
Object {
  "topics": Array [
    Object {
      "name": "other-topic",
      "partitions": Array [
        Object {
          "baseOffset": 0n,
          "errorCode": 0,
          "index": 999,
        },
      ],
    },
  ],
}
`;

exports[`Kafka API produce: send multiple message batches to one partiton 3`] = `"AAAAJwAAAAYAAAABAAtvdGhlci10b3BpYwAAAAEAAAPnAAAAAAAAAAAAAw=="`;

exports[`Kafka API produce: send multiple message batches to one partiton 4`] = `
Object {
  "topics": Array [
    Object {
      "name": "other-topic",
      "partitions": Array [
        Object {
          "baseOffset": 3n,
          "errorCode": 0,
          "index": 999,
        },
      ],
    },
  ],
}
`;

exports[`Kafka API produce: send one message to one partiton 1`] = `"AAAAJgAAAAUAAAABAAp0ZXN0LXRvcGljAAAAAQAAAU0AAAAAAAAAAAAA"`;

exports[`Kafka API produce: send one message to one partiton 2`] = `
Object {
  "topics": Array [
    Object {
      "name": "test-topic",
      "partitions": Array [
        Object {
          "baseOffset": 0n,
          "errorCode": 0,
          "index": 333,
        },
      ],
    },
  ],
}
`;
